### ISSUE TITLE
Chimeric contigs still present?

### ISSUE THREAD

**Issue by damioresegun** [malonge/RagTag#30]
Following on from another issue (Issue #29 ), I thought I would add this here. 

I'm finding that ragtag is not able to resolve chimeric contigs even after doing the ragtag correct protocol. You will see in the attached image different versions of a genome's chromosome AFTER scaffolding. 

The pertinent rows are the 3rd, 4th and 5th rows. The 3rd row is the RaGOO only process, the 4th RagTag only and 5th is RagTag then RaGOO process. 

As you can see RaGOO was able to resolve the chromosome somewhat while RagTag could not. Admittedly, this particular chromosome is weirdly tricky for unknown reasons but I found that in other chromosomes that the combination of both provided the best consensus scaffolding. Although the downside is that the use of RaGOO resulted in loss of some of what I'm assuming to be telomeric regions which RagTag often seems to get. 

Commands used: 
RaGOO: `ragoo.py isolate_masked.fasta reference.fasta.gz -t 24 -b`
RagTag: `ragtag correct reference.fasta isolate_masked.fasta -t 24`
             `ragtag scaffold reference.fasta isolated_masked.corrected.fasta -t 24`
Combined: same commands as above but ran RagTag first then RaGOO on the output of RagTag

![image](https://user-images.githubusercontent.com/33520829/95750320-def19700-0c94-11eb-8944-b6dd2a0af296.png)

**Comment by malonge** [comment]
Hi there,

Thanks for bringing this up. What organism is this? If I am interpreting this correctly, my guess is that most of the alignments are repetitive or otherwise filtered out by ragtag. Perhaps the first thing to check is the mapq score distribution in `query_against_ref.paf`. It might be that you have to adjust `-q`.

**Comment by damioresegun** [comment]
Well thats thing, its not being filtered out to Chr00 or something. Its being placed in an entire other chromosome. For example, entire sections of the above chromosome 2 were found in chromosome 7. I'll have a look at the mapq score though

**Comment by damioresegun** [comment]
Hello, 

Would the --debug be added for both correct and scaffold functions?

**Comment by malonge** [comment]
Yes you could add it for both.

**Comment by damioresegun** [comment]
Done both. Here is a subset of the .paf file showing the contigs which align against the reference chromosome 02 AND the other chromosomes those contigs align against

![image](https://user-images.githubusercontent.com/33520829/95892581-a62dec80-0d7e-11eb-8b67-592a486faf9b.png)

**Comment by damioresegun** [comment]
Oh also apologies forgot to add that this is a plasmodium genome. 24Mb; 14 chromosomes. Haploid as well

**Comment by malonge** [comment]
Hi there,

Thanks for sharing. Could you also share the merged alignments for contig 62 and 75 for correction? i.e.

```
grep "65" ragtag.correction.debug.merged.paf
```

**Comment by damioresegun** [comment]
Sure, here you go 

65      29388   874     7463    -       PKNH_08_v2      1888040 1183303 1190890 871     7956    22

**Comment by malonge** [comment]
Hi there,

Sorry I didn't ask for this earlier but could you also share the full ragtag commands that you used?

Thanks

**Comment by damioresegun** [comment]
Hello, I placed them in the original post on this issue. The only thing I added now is the --debug param

**Comment by malonge** [comment]
oh wow I am so sorry but I actually wanted to look at contig 62, not 65. Contig 75 gives me something to work with in the meantime.

Thanks

**Comment by damioresegun** [comment]
Haha no problems, here you go

62      157971  29474   134809  +       PKNH_07_v2      1522165 1087882 1200184 72965   112302  60
62      157971  24631   27740   -       PKNH_08_v2      1888040 1687950 1691206 95      3256    28

**Comment by malonge** [comment]
As I suspected the problem seems to be with the underlying alignments. When you align your original contigs to the reference for `ragtag correct`, the first ~1.3Mbp of contig 75 align exclusively to `PKNH_13_v2`. That yields the 75:0-1367571(+) contig. But then when you scaffold the broken contigs with `ragtag scaffold`, that contig aligns not only to `PKNH_13_v2` (as expected) but also to `PKNH_02_v2`. But why wouldn't it have aligned there during the correction?

If you are comfortable sharing you data, I can take a look to try to make sense of it. You also may try using Nucmer instead, especially since your genome is relatively small.

Thanks

**Comment by malonge** [comment]
You can send the data to malonge11@gmail.com.

One difference between ragtag and ragoo is that ragtag uses `-x asm5` and ragoo uses `-k19 -w19` minimap2 parameters. 

You can try the different parameters with ragtag like this:

`ragtag.py scaffold --mm2-params "-k19 -w19" ...`

Or you may also consider Nucmer, like I said above.

Thanks

**Comment by damioresegun** [comment]
Ahh right, I'll try those params for ragtag and see what happens on my end. 
I will certainly try nucmer as well. I'll keep you updated. 

Sent the data as well

**Comment by damioresegun** [comment]
Update: I've retried the data -- the same dataset I sent you using different tools and parameters. See screenshot below of one of my chromosomes. This chromosome has proven particularly problematic so I think its an apt example. 
To explain the rows: 
Row 1: Reference chromosome 2
Row 2: Another reference chromosome 2
Row 3: Prior Ragtag generated chromosome 2 -- generated July 2020. Default parameters for correct and scaffold functions
Row 4: Prior idea of carrying out RaGOO then Ragtag -- generated July 2020
Row 5: Newly generated RaGOO only chromosome 2
Row 6: Idea of using the output of the chimera_break step for my manual scaffolding rather than Ragoo scaffolded output
Row 7: RagOut -- another similar tool but takes SIGNIFICANTLY longer. Default parameters and concatenated unplaced sequences
Row 8: RagOut -- Default parameters without unplaced sequences
Row 9: Newly generated RagTag -- inferred gap size estimation off
Row 10: Newly generated RagTag -- inferred gap size estimation on
Row 11: New idea of running RaGOO first and using the output of the chimera_break as RagTag input as a way of retaining the smaller sequences which RaGOO would lose. 

![image](https://user-images.githubusercontent.com/33520829/96922543-e6d8e480-14a7-11eb-8bfe-cbaab7972bec.png)

I also tried changing the aligners used or the minimap parameters and in all cases, the outputs were worse then default. In many cases, adding >500kb of Ns! So decided not to carry them forward. Do you have any ideas on that front?

Generally, across these chromosomes, RagTag alone is just not working, the combination of the chimeric break of Ragoo and then Ragtag gets the best outcomes though I think RagOut is slightly better at retaining the smaller sequences. The speed of Ragoo and Ragtag completely blows it out the water though. 

In my particular use-case, it just seems like the chimeric detection/breaking within RagTag isn't working as expect but RaGOO is. 

Did you manage to find anything on your end?

**Comment by damioresegun** [comment]
Very late update on my end here. Apologies. Thanks to @malonge we kind of figured out what happened here. 
Basically it seems that its an issue with minimap2. For some reason there were repetitive alignments between the 'correct' and 'scaffold' functions so it ended up retaining sequences and placing them in the wrong scaffolds. 
@malonge suggested switching to the nucmer aligner which ended up working great. No issues downstream. At least none that I can see. 
The parameters for the nucmer I used were "-maxmatch -l 100 -c 500". 

So all sorted on this end. Closing this issue now.
